File: admin/src/types/fee-template.ts
export interface FeeTemplateExchangeItem {
  id?: number;
  from_currency: string;
  to_currency: string;
  fee_rate: string;
  min_fee: string;
  min_fee_currency: string;
}

export interface FeeTemplateCryptoWithdrawalItem {
  id?: number;
  currency: string;
  chain: string;
  fee_rate: string;
  fixed_fee: string;
}

export interface FeeTemplateFiatWithdrawalItem {
  id?: number;
  currency: string;
  transfer_type: string;
  fee_rate: string;
  fixed_fee: string;
}

export interface FeeTemplate {
  id: number;
  name: string;
  description?: string;
  is_default: boolean;
  exchange_items?: FeeTemplateExchangeItem[];
  crypto_withdrawal_items?: FeeTemplateCryptoWithdrawalItem[];
  fiat_withdrawal_items?: FeeTemplateFiatWithdrawalItem[];
  created_at: string;
  updated_at: string;
}

export interface CreateFeeTemplateRequest {
  name: string;
  description?: string;
  is_default?: boolean;
  exchange_items?: Omit<FeeTemplateExchangeItem, "id">[];
  crypto_withdrawal_items?: Omit<FeeTemplateCryptoWithdrawalItem, "id">[];
  fiat_withdrawal_items?: Omit<FeeTemplateFiatWithdrawalItem, "id">[];
}

export type UpdateFeeTemplateRequest = CreateFeeTemplateRequest;

File: backend/internal/dto/response/fee_template.go
package response

import "time"

type ExchangeItemResp struct {
	ID             uint64 `json:"id"`
	FromCurrency   string `json:"from_currency"`
	ToCurrency     string `json:"to_currency"`
	FeeRate        string `json:"fee_rate"`
	MinFee         string `json:"min_fee"`
	MinFeeCurrency string `json:"min_fee_currency"`
}

type CryptoWithdrawalItemResp struct {
	ID       uint64 `json:"id"`
	Currency string `json:"currency"`
	Chain    string `json:"chain"`
	FeeRate  string `json:"fee_rate"`
	FixedFee string `json:"fixed_fee"`
}

type FiatWithdrawalItemResp struct {
	ID           uint64 `json:"id"`
	Currency     string `json:"currency"`
	TransferType string `json:"transfer_type"`
	FeeRate      string `json:"fee_rate"`
	FixedFee     string `json:"fixed_fee"`
}

type FeeTemplateListItem struct {
	ID          uint64    `json:"id"`
	Name        string    `json:"name"`
	Description *string   `json:"description,omitempty"`
	IsDefault   bool      `json:"is_default"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type FeeTemplateListResp struct {
	Templates []FeeTemplateListItem `json:"templates"`
}

type FeeTemplateDetailResp struct {
	ID                                 uint64                     `json:"id"`
	Name                               string                     `json:"name"`
	Description                        *string                    `json:"description,omitempty"`
	IsDefault                          bool                       `json:"is_default"`
	ExchangeFeeDeductionMethod         string                     `json:"exchange_fee_deduction_method"`
	CryptoWithdrawalFeeDeductionMethod string                     `json:"crypto_withdrawal_fee_deduction_method"`
	FiatWithdrawalFeeDeductionMethod   string                     `json:"fiat_withdrawal_fee_deduction_method"`
	ExchangeItems                      []ExchangeItemResp         `json:"exchange_items"`
	CryptoWithdrawalItems              []CryptoWithdrawalItemResp `json:"crypto_withdrawal_items"`
	FiatWithdrawalItems                []FiatWithdrawalItemResp   `json:"fiat_withdrawal_items"`
	CreatedAt                          time.Time                  `json:"created_at"`
	UpdatedAt                          time.Time                  `json:"updated_at"`
}
